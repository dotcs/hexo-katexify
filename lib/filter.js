var katex = require('katex');

function filter(data) {
  var displayRegExp = new RegExp(/\$\$[^$]*\$\$/, 'mg');
  data.content = data.content.replace(displayRegExp, function(s) {
    return katex.renderToString(s.slice(2, -2), { displayMode: true });
  });

  var regexp = new RegExp(/\$[^$]*\$/, 'mg');
  data.content = data.content.replace(regexp, function(s) {
    return katex.renderToString(s.slice(1, -1));
  });
  return data;
}

module.exports = filter;
