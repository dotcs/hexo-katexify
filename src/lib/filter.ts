import { renderToString } from "katex";
import { unescape } from "he";

export function filter(data: { content: string }) {
  const displayRegExp = new RegExp("\\$\\$[^\\$]*\\$\\$", "mg");
  data.content = data.content.replace(displayRegExp, (s: string) => {
    return renderToString(unescape(s).slice(2, -2).trim(), {
      displayMode: true
    });
  });

  const regexp = new RegExp("\\$([^\\$]*)\\$", "mg");
  data.content = data.content.replace(regexp, (s: string) => {
    return renderToString(unescape(s).slice(1, -1).trim());
  });
  return data;
}
