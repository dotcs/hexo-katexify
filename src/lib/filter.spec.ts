import { filter } from "./filter";

describe("filter", () => {
  it("should not do anything if content does not contain katex formula", () => {
    const content = `
# This is some content
<!-- ignore this comment -->
<h2 class="to-be-ignored">Please ignore me</h2>
Here we have do no nothing. I repeat: Nothing!
    `;
    expect(filter({ content })).toEqual({ content });
  });

  it("should return the same object reference", () => {
    const content = `
# This is some content
<!-- ignore this comment -->
<h2 class="to-be-ignored">Please ignore me</h2>
Here we have do no nothing. I repeat: Nothing!
    `;
    const obj = { content };
    expect(filter(obj)).toBe(obj);
  });

  it("should find inline formula and replace it", () => {
    const content = `
# This is some content
<!-- ignore this comment -->
<h2 class="to-be-ignored">Please ignore me</h2>
This is a formula: $E = mc^2$.
    `;
    const { content: responseContent } = filter({ content });
    expect(responseContent).not.toEqual(content);
    expect(responseContent).toContain("This is some content");
    expect(responseContent).toContain("This is a formula");
    expect(responseContent).not.toContain("katex-display");
    expect(responseContent).not.toContain("$E = mc^2$");
    expect(responseContent).not.toContain("$");
    expect(responseContent).toMatchSnapshot();
  });

  it("should find display formula and replace it", () => {
    const content = `
# This is some content
<!-- ignore this comment -->
<h2 class="to-be-ignored">Please ignore me</h2>
This is a formula: $$E = mc^2$$.
    `;
    const { content: responseContent } = filter({ content });
    expect(responseContent).not.toEqual(content);
    expect(responseContent).toContain("This is some content");
    expect(responseContent).toContain("This is a formula");
    expect(responseContent).toContain("katex-display");
    expect(responseContent).not.toContain("$$E = mc^2$$");
    expect(responseContent).not.toContain("$");
    expect(responseContent).toMatchSnapshot();
  });
});
