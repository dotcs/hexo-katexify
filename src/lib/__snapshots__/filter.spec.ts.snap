// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter should find display formula and replace it 1`] = `
"
# This is some content
<!-- ignore this comment -->
<h2 class=\\"to-be-ignored\\">Please ignore me</h2>
This is a formula: <span class=\\"katex-display\\"><span class=\\"katex\\"><span class=\\"katex-mathml\\"><math><semantics><mrow><mi>E</mi><mo>=</mo><mi>m</mi><msup><mi>c</mi><mn>2</mn></msup></mrow><annotation encoding=\\"application/x-tex\\">E = mc^2</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"strut\\" style=\\"height:0.8641079999999999em;\\"></span><span class=\\"strut bottom\\" style=\\"height:0.8641079999999999em;vertical-align:0em;\\"></span><span class=\\"base displaystyle textstyle uncramped\\"><span class=\\"mord mathit\\" style=\\"margin-right:0.05764em;\\">E</span><span class=\\"mrel\\">=</span><span class=\\"mord mathit\\">m</span><span class=\\"mord\\"><span class=\\"mord mathit\\">c</span><span class=\\"msupsub\\"><span class=\\"vlist\\"><span style=\\"top:-0.413em;margin-right:0.05em;\\"><span class=\\"fontsize-ensurer reset-size5 size5\\"><span style=\\"font-size:0em;\\">​</span></span><span class=\\"reset-textstyle scriptstyle uncramped mtight\\"><span class=\\"mord mathrm mtight\\">2</span></span></span><span class=\\"baseline-fix\\"><span class=\\"fontsize-ensurer reset-size5 size5\\"><span style=\\"font-size:0em;\\">​</span></span>​</span></span></span></span></span></span></span></span>.
    "
`;

exports[`filter should find inline formula and replace it 1`] = `
"
# This is some content
<!-- ignore this comment -->
<h2 class=\\"to-be-ignored\\">Please ignore me</h2>
This is a formula: <span class=\\"katex\\"><span class=\\"katex-mathml\\"><math><semantics><mrow><mi>E</mi><mo>=</mo><mi>m</mi><msup><mi>c</mi><mn>2</mn></msup></mrow><annotation encoding=\\"application/x-tex\\">E = mc^2</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"strut\\" style=\\"height:0.8141079999999999em;\\"></span><span class=\\"strut bottom\\" style=\\"height:0.8141079999999999em;vertical-align:0em;\\"></span><span class=\\"base textstyle uncramped\\"><span class=\\"mord mathit\\" style=\\"margin-right:0.05764em;\\">E</span><span class=\\"mrel\\">=</span><span class=\\"mord mathit\\">m</span><span class=\\"mord\\"><span class=\\"mord mathit\\">c</span><span class=\\"msupsub\\"><span class=\\"vlist\\"><span style=\\"top:-0.363em;margin-right:0.05em;\\"><span class=\\"fontsize-ensurer reset-size5 size5\\"><span style=\\"font-size:0em;\\">​</span></span><span class=\\"reset-textstyle scriptstyle uncramped mtight\\"><span class=\\"mord mathrm mtight\\">2</span></span></span><span class=\\"baseline-fix\\"><span class=\\"fontsize-ensurer reset-size5 size5\\"><span style=\\"font-size:0em;\\">​</span></span>​</span></span></span></span></span></span></span>.
    "
`;
